Refactoring to be able to improve information efficiency

felixbauckholt opened the pull request:
Commit 285ae7a does a lot of refactoring, the subsequent commits are all changes that I think are made much easier by the refactoring. Most of the strategy changes in the subsequent commits were discussed in #6.

felixbauckholt committed:
Switch around ordering of question combining

This has the advantage that for some ModulusInformation object m and
some modulus,

```
a = m.split(modulus);
m.cast_down(x);
```

is equivalent to

```
m.cast_down(x*modulus);
a = m.split(modulus);
```

This means that when using a ModulusInformation object to answer
questions, we can answer the first question without needing to know how
many questions there are.

felixbauckholt committed:
Refactor out a "public information object"

One important change is that now, when deciding which questions to ask, they can see the answer to the last question before asking the next one.

Some design choices:
- Questions now take a BoardState instead of an OwnedGameView.
- When deciding which questions to ask (in ask_questions), we get an immutable public information object
  (representing the public information before any questions were asked), and a mutable HandInfo<CardPossibilityTable>
  that gets updated as we ask questions. That HandInfo<CardPossibilityTable> was copied instead of taken.
- In ask_questions, we also get some &mut u32 representing "info_remaining" that gets updated for us.
  This will later allow for cases where "info_remaining" depends on the answers to previous questions.
- Both get_hint_sum and update_from_hint_sum change the public information object. If you want to compute the
  hint sum but aren't sure if you actually want to give the hint, you'll have to clone the public information
  object!
- Over time, in the code to decide on a move, we'll be able to build an increasingly complicated tree of
  "public information object operations" that will have to be matched exactly in the code to update on a move.
  In order to make this less scary, I moved most of the code into
  "decide_wrapped" and "update_wrapped". If the call to update_wrapped
  (for the player who just made the move) changes the public information
  object in different ways than the previous call to decide_wrapped, we
  detect this and panic.

This commit should be purely refactoring; all changes to win-rates are
due to bugs.

felixbauckholt committed:
Slight changes to be more information-efficient

- If a player discards because nobody else needed a hint, that
  information is "transmitted" before they transmit other info through
  hat stuff.
- The "card possibility partition" questions now take into account info
  learned through earlier questions.

felixbauckholt committed:
Modulus magic!

Suppose we have `total_information` choices, and we first use them to
encode the answer `x` to a question with `m` answers. That answer is encoded
by the choice we take modulo `m`.

How much "information" do we have left? That depends on the number of
numbers less than `total_information` that are equal to `x` modulo `m`.
Depending on the value of `x`, this is either
`floor(total_information/m)` or `floor(total_information/m) + 1`.

We now use all of this information as opposed to just
`floor(total_information/m)`, at the cost of making our math not a lot
more complicated but pretty confusing.

WuTheFWasThat commented:
Thanks a lot for your work on this!

Haven't looked in detail yet, but design decisions in the commit message mostly look good!  One kinda minor additional design thing:  Should we switch to just only asking one question at a time?  It's strictly better right?  And might simplify things too (on the other hand it might make it more awkward to know when to do the additive question infos)

felixbauckholt commented:
I'm glad it looks reasonable!

What do you mean by asking one question question at a time?

If you mean replacing `ask_questions` (with the callback) by a function that gets called several times and returns a question instead of having to call the callback, yes I agree that would look much nicer! I wouldn't want to implement our current strategy this way, since we'd have to implement a somewhat silly state machine (for five players: "if we don't have a playable card or we previously asked an IsPlayable question, ask an IsPlayable question, otherwise ask a `CardPossibilityPartition` question"), and also put that state somewhere. But if we want to change the strategy to get rid of that state (and we don't think we'll want to reintroduce it), I'd want to make that change!

If you mean getting rid of the `AdditiveComboQuestion`s, no that wouldn't be strictly better: If you have 5 "information" and you want to emulate an `AdditiveComboQuestion` of four binary questions by asking them one by one, you'd need some way to ask the first question such that the amount of info remaining is 1 if the answer was 0 and 4 if the answer was 1. That's of course possible somehow, but I don't see how it's easier than using `AdditiveComboQuestion`.

WuTheFWasThat commented:
I meant the first, and yep, I didn't explain well, but I was worried essentially about having to implement silly state machine.  And yeah, AdditiveComboQuestion would still be a single question, as it's fundamentally different from asking the questions separately.

Regarding strategy change:  I'm pretty okay with deviating from the current strategy, as I'm not sure it makes that much sense.  It seems a bit nicer to be able to experiment with the optimal question at any state, instead of the optimal batch of questions for a given hint (it seems like potentially a real tradeoff, but actually in the worst case we could have MultiplicativeIndependentComboQuestion and have the current thing lol).

felixbauckholt committed:
Improve ask_question strategy

I also replaced the callback architecture behind
`PublicInformation.ask_questions()` with a simpler method
`PublicInformation.ask_question()` that gets called repeatedly.

To make all float-based sorts easier, I used the `float-ord` package.
I also used it to clean up some of the sorting in `decide_wrapped()`.

WuTheFWasThat commented:
i got a chance to skim this over, looks great.  actually a really big improvement too!  do you know where most of it comes from?

felixbauckholt committed:
Merge branch 'master' into information-efficiency

felixbauckholt commented:
Uhh I don't think I have any intuitions that are going to be more accurate than the changes to README.md that you can see in each individual commit. One thing that might be important to keep in mind is that since we're reusing seeds, things behave like in a "matched" experiment, in the sense that while the reported win rates are some estimate of the "true" win rates (with the reported standard errors), the differences between reported win rates may be a more accurate estimate of the differences between "true" win rates.

Regarding the lack of improvement from bd56380121a5c2d9ea55e26b669ca65e3fdca8f6, I wanted to see if its change interacts with the later changes, so as an experiment, I reverted that commit here: 57793ebaa5ef6360f45c188f366571c3d6ea574b. There doesn't seem to be any interaction (it's hard to tell because this commit's effect is so tiny), so it should be safe to cherry-pick out this commit if you want.
By the way, I still don't have a good intuition why the change wouldn't have a greater effect.

EDIT: I also don't know where the win rate regression in 493631dad01ff037046a5c6f2854bbf2c0020942 comes from.

WuTheFWasThat commented:
oh man, so much of the change comes from 051ac7a097682a2278dea573220cc27659e3c03e!  I wouldn't worry too much about lack of improvement or regression on bd56380 and 493631d.  They seem within noise, and presumably changing around stuff is affecting how the seeds are used?

WuTheFWasThat commented on src/strategies/information.rs:
```diff
@@ -892,12 +895,17 @@ impl InformationPlayerStrategy {
+        if public_useless_indices.len() > 0 {
+            let info_amt = public_useless_indices.len() as u32;
+            if let Some(private_useless_index) = private_useless_index {
+                let num_states = info_amt + 2;
```
hmm, i might be confused, but shouldn't it just always be 2?  why also use info_amt?

felixbauckholt replied:
Ohh yes I should probably have elaborated on that better, sorry!
So we'll have a `1/num_states` chance of discarding the privately known to be useless card, in which case we transmit `log2(num_states)` bits of information. All else being equal, we thus want to maximize `log2(num_states)/num_states`, which we get by setting `num_states = 3`.

But if `info_amt != 1`, all else is not equal, since if we don't choose to play the privately known useless card, we get to transmit `log2(info_amt)` bits of information. Therefore, we want to increase `num_states` so that we're less likely to play the privately known card (and learn more if we do).

Actually, writing this got me to plug this into Wolfram Alpha, and there's a nice optimum: I should have set `num_states` to some rounding of `e * info_amt`! I don't know why I didn't have that idea earlier.

WuTheFWasThat replied:
my (weak) intuition is actually that *always* playing the privately known useless card will do better than all of this, in the limit

advantages:
- you keep more publicly known discards, which lead to more info transmission later on (although maybe this mainly happens in endgame)
- when you discard a publicly known card, other players learn that you don't have a privately known dead card.  that's a lot of information in the (admittedly) rare case it happens.  the information is typically of the form "i have a red 4", though, which we currently can't make use of

disadvantage:
- when the card is discarded you dont transmit any additional question-based info

WuTheFWasThat commented on src/strategies/information.rs:
```diff
@@ -892,12 +895,17 @@ impl InformationPlayerStrategy {
+                    return TurnChoice::Discard(*private_useless_index);
+                }
+            }
+            let info = public_info.get_hat_sum(info_amt, view);
```
i think one problem is we're now in this case even if public_useless_indices.len() == 1.  previously the idea is it's better to hold onto these discards since once you have 2 of them you get a free bit of info

felixbauckholt replied:
Ohh, I didn't consider that at all! A special case makes sense here (I think, but I'm not too sure) since it's the only case in which there aren't decreasing returns to having public useless cards: If we have privately known useless cards and discard, the number of bits we get is

| public useless cards | bits without this trick | bits with this trick (roughly) |
|------|--------|-------|
| 0 | 0 | 0 |
| 1 | 0 | 0.5 |
| 2 | 1 | 1.3 |
| 3 | 1.6 | 1.8 |
| 4 | 2 | 2.1 |

I'll report back if the special case ends up helping.

felixbauckholt replied:
Ok, the results are:

Reference (current tip unless I made a mistake):

|         |   2p    |   3p    |   4p    |   5p    |
|---------|------------------|------------------|------------------|------------------|
| cheat   | 24.8594 ± 0.0036 | 24.9785 ± 0.0012 | 24.9720 ± 0.0014 | 24.9557 ± 0.0018 |
|         | 90.59 ± 0.21 % | 98.17 ± 0.09 % | 97.76 ± 0.10 % | 96.42 ± 0.13 % |
| info    | 22.5239 ± 0.0125 | 24.7952 ± 0.0039 | 24.9353 ± 0.0022 | 24.9221 ± 0.0024 |
|         | 12.63 ± 0.23 % | 84.54 ± 0.26 % | 95.01 ± 0.15 % | 94.01 ± 0.17 % |

With `num_states = round(e * info_amt)`, without special case:

|         |   2p    |   3p    |   4p    |   5p    |
|---------|------------------|------------------|------------------|------------------|
| cheat   | 24.8594 ± 0.0036 | 24.9785 ± 0.0012 | 24.9720 ± 0.0014 | 24.9557 ± 0.0018 |
|         | 90.59 ± 0.21 % | 98.17 ± 0.09 % | 97.76 ± 0.10 % | 96.42 ± 0.13 % |
| info    | 22.5238 ± 0.0125 | 24.7950 ± 0.0039 | 24.9353 ± 0.0022 | 24.9221 ± 0.0024 |
|         | 12.62 ± 0.23 % | 84.54 ± 0.26 % | 95.01 ± 0.15 % | 94.02 ± 0.17 % |

With `num_states = round(e * info_amt)`, except if there's exactly one public useless card, in which case we discard the private one:

|         |   2p    |   3p    |   4p    |   5p    |
|---------|------------------|------------------|------------------|------------------|
| cheat   | 24.8594 ± 0.0036 | 24.9785 ± 0.0012 | 24.9720 ± 0.0014 | 24.9557 ± 0.0018 |
|         | 90.59 ± 0.21 % | 98.17 ± 0.09 % | 97.76 ± 0.10 % | 96.42 ± 0.13 % |
| info    | 22.5208 ± 0.0125 | 24.7945 ± 0.0039 | 24.9353 ± 0.0022 | 24.9221 ± 0.0024 |
|         | 12.58 ± 0.23 % | 84.47 ± 0.26 % | 95.02 ± 0.15 % | 94.01 ± 0.17 % |

So the special case doesn't appear to help, but I think the real takeaway is that the differences are all tiny, presumably because (I'd guess) privately, but not publicly known useless cards are just too rare to matter. So I'm totally fine with any version getting merged!

WuTheFWasThat replied:
cool, i think we should probably just not add it then, for now; given fixed performance, simpler strategies seem better :)

WuTheFWasThat commented:
i'm fine with all these changes besides bd56380121a5c2d9ea55e26b669ca65e3fdca8f6

felixbauckholt commented:
Ok, I rebased out bd56380121a5c2d9ea55e26b669ca65e3fdca8f6. Let me know if you need anything else!

WuTheFWasThat commented:
Nope!
On Fri, Mar 15, 2019 at 9:40 AM Felix Bauckholt <notifications@github.com>
wrote:

> Ok, I rebased out bd56380
> <https://github.com/WuTheFWasThat/hanabi.rs/commit/bd56380121a5c2d9ea55e26b669ca65e3fdca8f6>.
> Let me know if you need anything else!
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/WuTheFWasThat/hanabi.rs/pull/7#issuecomment-473358102>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABaT4KLZwzzEABCfIDyjTDZMyrZ2ZPX9ks5vW8z3gaJpZM4bg3p5>
> .
>

felixbauckholt commented:
> Nope!

I'll admit I'm confused: Do you want to press that green button now? Do you want me to press it?

WuTheFWasThat commented:
Go ahead and merge!  Although there appears to be conflicts..
